### PR TITLE
Try reducing specificity of global styles selectors only.

### PR DESCRIPTION
**Pull request by tellthemachines**
## What?

Split out from #57841. This part handles only the global styles selectors: block, element and feature level. The layout selectors will follow in another PR, but I think, from my local testing, that this change works pretty well by itself. 

## Testing Instructions

1. Add global styles to all the blocks and check that they are applied as expected in both the editor and the front end.

### CODE REVIEW

**Comment by MaggieCabrera** [comment]
On a first quick glance this is looking good on my testing. Are we planning on changing specificity to the block's CSS or is that something we want to keep as is? I'm looking at the block's stylesheets to see what could be overriding theme styling via theme.json and I don't find much, so it might not be a big deal. I'll come back with more testing.

**Comment by MaggieCabrera** [comment]
> On a first quick glance this is looking good on my testing. Are we planning on changing specificity to the block's CSS or is that something we want to keep as is? I'm looking at the block's stylesheets to see what could be overriding theme styling via theme.json and I don't find much, so it might not be a big deal. I'll come back with more testing.

Ok, delving deeper into the blocks' stylesheets I see that we have already covered our based in terms of what can be overrideable via GS. We need to think that in the future some of the things that can't be styled via GS will be and we will need to go back to the blocks' styles to ensure that specificity war is being won, but that shouldn't block this PR. Maybe we need to document that somewhere.

**Review by MaggieCabrera — COMMENTED** [review]
There are some weird `:where` nesting when we have some elements inside blocks on theme.json. The final output seems to be working fine and the specificity is 0 like expected, but the CSS is a bit ugly I suppose. Not a blocker but I guess worth commenting. 

The case in my example on TT4 would be:

```
"core/post-title": {
	"elements": {
		"link": {
			"typography": {
				"textDecoration": "none"
			}
		}
	}
},
```

Turning into:

```
:where(.wp-block-post-title a:where(:not(.wp-element-button))) {
  text-decoration: none;
}
```


Where in trunk it's just:

```
.wp-block-post-title a:where(:not(.wp-element-button)) {
  text-decoration: none;
}
```

**Comment by tellthemachines** [comment]
Thanks for reviewing and testing @MaggieCabrera !

> There are some weird `:where` nesting when we have some elements inside blocks on theme.json. The final output seems to be working fine and the specificity is 0 like expected, but the CSS is a bit ugly I suppose.

I ended up not removing the `:where` from the element selector after [this conversation](https://github.com/WordPress/gutenberg/pull/57841/files#r1495265568) on the other PR. We could still try it, but given the current selector works, it might be better to tackle that separately as a follow-up in order to avoid blocking this PR 😅 

> Are we planning on changing specificity to the block's CSS or is that something we want to keep as is?

It would be good to ensure minimal specificity for all block-library styles as that's the general direction we've been heading in for a few years now. Again, that could be done as a follow-up.

Also, as has already been commented in #57841, CSS cascade layers could be a great candidate for solving this problem in one go, but an exploration is needed to see if it's actually a viable solution for us right now. In the meantime, reducing global styles specificity is the main blocker for the extension of block style variations, so we're proceeding down this safer route for now.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for continuing to split up and work through #57841 👍 

It looks like we're missing the `:where` around global styles for blocks that use custom feature-level selectors e.g Image block and borders.

The [required change](https://github.com/WordPress/gutenberg/pull/57841/files#diff-52acb55f718dce035b748387f60ef068cd3769674b553584f7df6f790e9fa84eL842) is in the original PR and works when applied here. Once applied though the unit tests for the `useGlobalStylesOutput` hook will need updating as per #57841 too.

Screenshot 2024-03-25 at 1 05 42 PM

I've mostly tested via the site editor's style book so far and from that side most blocks are working as expected.

**Comment by tellthemachines** [comment]
> It looks like we're missing the `:where` around global styles for blocks that use custom feature-level selectors e.g Image block and borders.

Well spotted! Updated now. Thanks for reviewing!

**Comment by MaggieCabrera** [comment]
Sorry I missed some of the comments on the original PR! yeah, none of my comments are blockers, it seems like you've covered all the bases

**Review by aaronrobertshaw — APPROVED** [review]
I've rebased this one to resolve the conflicts with the theme.json unit tests after the recent switch from `body` to `:root` for custom CSS properties (https://github.com/WordPress/gutenberg/pull/42084).

Following that I've also given this another smoke test and it's testing well for me. Using the site editor's style book, all the global styles I tweaked were applied to the block. Those styles were also then reflected on the frontend still.

The prior issue with the feature selectors has been resolved as well.

Screenshot 2024-03-26 at 4 12 57 PM

I think this is in reasonable shape so I'll give it a tentative approval. Given the broad nature of the changes, it might pay to get an extra set of eyes from @MaggieCabrera before hitting merge.

File: lib/class-wp-theme-json-gutenberg.php
Comment (MaggieCabrera) [inline lib/class-wp-theme-json-gutenberg.php:2686]: do we want to apply this to duotone selectors too?

File: packages/block-editor/src/components/iframe/index.js
Comment (MaggieCabrera) [inline packages/block-editor/src/components/iframe/index.js:247]: Can confirm that this doesn't break the zoomed-out mode 👍

**Review by MaggieCabrera — APPROVED** [review]
This is looking good to me too, let's bring this in!

**Comment by tellthemachines** [comment]
Awesome, thanks for reviewing folks!

**Pull request merged**
